File: system-tests/http-redirect-test/src/test/java/com/teamscale/client/RedirectMockServer.java
package com.teamscale.client;

import spark.Request;
import spark.Response;
import spark.Service;

import static javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR;

/**
 * Mocks a redirect server: redirects all requests to another server.
 */
public class RedirectMockServer {

	private final Service service;

	public RedirectMockServer(int port, int redirectTo) {
		service = Service.ignite();
		service.port(port);
		service.post("/*", (Request request, Response response) -> {
			String url = request.url();
			url = url.replace(Integer.toString(port), Integer.toString(redirectTo));
			response.redirect(url, 307);
			return response;
		});
		service.exception(Exception.class, (Exception exception, Request request, Response response) -> {
			response.status(SC_INTERNAL_SERVER_ERROR);
			response.body("Exception: " + exception.getMessage());
		});
		service.awaitInitialization();
	}


	/**
	 * Shuts down the mock server and waits for it to be stopped.
	 */
	public void shutdown() {
		service.stop();
		service.awaitStop();
	}

}

File: system-tests/http-redirect-test/src/test/kotlin/com/teamscale/client/RedirectMockServer.kt
package com.teamscale.client

import spark.*
import javax.servlet.http.HttpServletResponse

/**
 * Mocks a redirect server: redirects all requests to another server.
 */
class RedirectMockServer(port: Int, redirectTo: Int) {
	private val service: Service = Service.ignite()

	init {
		service.port(port)
		service.post("/*") { request, response ->
			var url = "${request.url()}?${request.queryString()}"
			url = url.replace(port.toString(), redirectTo.toString())
			response.redirect(url, 307)
			response
		}
		service.exception<Exception>(Exception::class.java) { exception, _, response ->
			response.status(HttpServletResponse.SC_INTERNAL_SERVER_ERROR)
			response.body("Exception: ${exception.message}")
		}
		service.awaitInitialization()
	}

	/**
	 * Shuts down the mock server and waits for it to be stopped.
	 */
	fun shutdown() {
		service.stop()
		service.awaitStop()
	}
}
